src/hir/combiner.rs

```diff
@@ -0,0 +1,164 @@
+/// by appending the function nodes to the root contract node.
+pub struct Combiner;
+
+impl Combiner {
+    /// Creates a new combiner.
+    #[must_use]
+    pub fn new() -> Self {
+        Combiner {}
+    }
+
+    /// Create a new error with the given span and error type.
+    fn error(&self, text: String, span: Span, kind: ErrorKind) -> Error {
+        Error {
+            kind,
+            text,
+            span,
+        }
+    }
+
+    /// Combines the translated HIRs into a single HIR. HIRs are merged by
+    /// iterating over each HIR and merging their children into the contract
+    /// definition of the first HIR, while verifying the contract identifiers
+    /// match and filtering out duplicate modifiers.
+    ///
+    /// This function is called after the ASTs are translated to HIR.
+    pub fn combine(&self, hirs: &Vec<Hir>) -> Result<Hir> {
+        let mut root: Root = Root::default();
+        let mut contract_definition = &mut ContractDefinition::default();
+        let mut added_modifiers = HashSet::new();
+
+        for hir in hirs {
+            match hir {
+                Hir::Root(r) => {
+                    for child in &r.children {
+                        match child {
+                            // check the ith HIR's identifier matches the accumulated ContractDefinition identifier
+                            // all the ContractDefinitions should be merged into a single child ContractDefinition with the same identifier
+                            Hir::ContractDefinition(contract_def) => {
+                                if contract_definition.identifier.is_empty() {
+                                    let mut child_contract = contract_def.clone();
+                                    child_contract.identifier = get_contract_name_from_identifier(&contract_def.identifier);
+                                    root.children.push(Hir::ContractDefinition(child_contract));
+                                    contract_definition = match &mut root.children[0] {
+                                        Hir::ContractDefinition(contract) => contract,
+                                        _ => unreachable!(),
+                                    }
+                                } else {
+                                    let text = contract_def.identifier.clone();
+                                    let identifier = get_contract_name_from_identifier(&text);
+                                    let accumulated_identifier = contract_definition.identifier.clone();
+                                    if identifier != accumulated_identifier {
+                                        let span = Span::default(); // @follow-up - how can we get the span from the HIR? Is it even necessary? This would be easier to do with verification of the AST. One option is to use the index of the HIR in the vector of HIRs since we know the identifier is the start of a given tree.
```

Comment:
Yeah, this is a good idea, but I'm hesitant to commit until I think more about it. We can report with the empty span for now.